packages/compiler-cli/src/ngtsc/metadata/src/field_mapping.ts

```diff
@@ -0,0 +1,185 @@
+ */
+export interface InputOrOutput {
+  /**
+   * The field name on the component or directive instance for this input or output.
+   */
+  fieldName: string;
+
+  /**
+   * The property name used to bind this input or output in an Angular template.
+   */
+  propertyName: string;
+}
+
+/**
+ * A mapping of field and property names, such as that for a particular directive or component.
+ *
+ * A single field name has exactly one input/output (and therefore one property name) associated
+ * with it, but the same property name may be shared across many field names.
+ *
+ * Allows bidirectional querying of the mapping - looking up all inputs/outputs with a given
+ * property name, or mapping from a specific field to its property name.
+ */
+export class FieldPropertyMapping implements InputOutputPropertySet {
+  /**
+   * Mapping from field names to the single `InputOrOutput` for that field.
+   */
+  private forwardMap: Map<FieldName, InputOrOutput>;
+
+  /**
+   * Mapping from property names to one or more `InputOrOutput`s which share that name.
+   */
+  private reverseMap: Map<PropertyName, InputOrOutput[]>;
+
+  private constructor(forwardMap: Map<FieldName, InputOrOutput>) {
+    this.forwardMap = forwardMap;
+    this.reverseMap = reverseMapFromForwardMap(forwardMap);
+  }
+
+  /**
+   * Construct a `FieldPropertyMapping` with no entries.
+   */
+  static empty(): FieldPropertyMapping {
+    return new FieldPropertyMapping(new Map());
+  }
+
+  /**
+   * Construct a `FieldPropertyMapping` from a primitive JS object which maps field names to either
+   * property names or an array that contains both names, which is used in on-disk metadata formats
+   * (e.g. in .d.ts files).
+   */
+  static fromFieldMappedObject(
+      obj: {[fieldName: string]: string|[/* propertyName */ string, /* fieldName */ string]}):
+      FieldPropertyMapping {
+    const forwardMap = new Map<FieldName, InputOrOutput>();
+
+    for (const fieldName of Object.keys(obj)) {
+      const value = obj[fieldName];
+      const propertyName = Array.isArray(value) ? value[0] : value;
+      const inputOrOutput: InputOrOutput = {fieldName, propertyName};
+      forwardMap.set(fieldName, inputOrOutput);
+    }
+
+    return new FieldPropertyMapping(forwardMap);
+  }
+
+  /**
+   * Merge two mappings into one, with fields from `b` taking precedence over fields from `a`.
+   */
+  static merge(a: FieldPropertyMapping, b: FieldPropertyMapping): FieldPropertyMapping {
+    const forwardMap = new Map<FieldName, InputOrOutput>(a.forwardMap.entries());
+    for (const [fieldName, inputOrOutput] of b.forwardMap) {
+      forwardMap.set(fieldName, inputOrOutput);
+    }
+
+    return new FieldPropertyMapping(forwardMap);
+  }
+
+  /**
+   * All field names mapped in this mapping.
+   */
+  get fieldNames(): string[] {
+    return Array.from(this.forwardMap.keys());
+  }
+
+  /**
+   * All property names mapped in this mapping.
+   */
+  get propertyNames(): string[] {
```

Comment:
```suggestion
  get propertyNames(): PropertyName[] {
```
There are many more places in this file where I think this change could be made?